pkg/dyninst/decode/types.go

```diff
@@ -0,0 +1,955 @@
+		}
+		if used > int64(totalElementsEncoded) {
+			if err := writeTokens(enc,
+				notCapturedReason,
+				notCapturedReasonPruned,
+			); err != nil {
+				return err
+			}
+		}
+	} else {
+		// This is a 'large' swiss map where there are multiple groups of data/control words
+		// We need to collect the data items for the table pointers first.
+		tablePtrSliceDataItemPtr, ok := d.dataItems[typeAndAddr{
+			irType: uint32(s.TablePtrSliceType.GetID()),
+			addr:   dirPtr,
+		}]
+		if !ok {
+			return writeTokens(enc,
+				jsontext.EndArray,
+				notCapturedReason,
+				notCapturedReasonDepth,
+			)
+		}
+		tablePtrSliceDataItem, ok := d.dataItems[typeAndAddr{
+			irType: tablePtrSliceDataItemPtr.Header().Type,
+			addr:   tablePtrSliceDataItemPtr.Header().Address,
+		}]
+		if !ok {
+			return writeTokens(enc,
+				jsontext.EndArray,
+				notCapturedReason,
+				notCapturedReasonDepth,
+			)
+		}
+		totalElementsEncoded, err := s.encodeSwissMapTables(d, enc, tablePtrSliceDataItem)
+		if err != nil {
+			return err
+		}
+		if used > int64(totalElementsEncoded) {
+			if err := writeTokens(enc,
+				notCapturedReason,
+				notCapturedReasonPruned,
+			); err != nil {
+				return err
+			}
+		}
+	}
+	return writeTokens(enc, jsontext.EndArray)
+}
+
+func (s *goSwissMapGroupsType) irType() ir.Type { return (*ir.GoSwissMapGroupsType)(s) }
+func (s *goSwissMapGroupsType) encodeValueFields(
+	_ *Decoder,
+	enc *jsontext.Encoder,
+	_ []byte,
+) error {
+	return writeTokens(enc,
+		notCapturedReason,
+		notCapturedReasonUnimplemented,
+	)
+}
+
+func (v *voidPointerType) irType() ir.Type { return (*ir.VoidPointerType)(v) }
+func (v *voidPointerType) encodeValueFields(
+	_ *Decoder,
+	enc *jsontext.Encoder,
+	data []byte,
+) error {
+	if len(data) != 8 {
+		return errors.New("passed data not long enough for void pointer")
+	}
+	return writeTokens(enc,
+		jsontext.String("address"),
+		jsontext.String("0x"+strconv.FormatUint(binary.NativeEndian.Uint64(data), 16)),
+	)
+}
+
+func (p *pointerType) irType() ir.Type { return (*ir.PointerType)(p) }
+func (p *pointerType) encodeValueFields(
+	d *Decoder,
+	enc *jsontext.Encoder,
+	data []byte,
+) error {
+	if len(data) < 8 {
+		return errors.New("passed data not long enough for pointer: need 8 bytes")
+	}
+	addr := binary.NativeEndian.Uint64(data)
+	pointeeKey := typeAndAddr{
+		irType: uint32(p.Pointee.GetID()),
+		addr:   addr,
+	}
+	if pointeeKey.addr == 0 {
+		if err := writeTokens(enc,
+			jsontext.String("isNull"),
+			jsontext.Bool(true),
+		); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	var address uint64
+	pointedValue, dataItemExists := d.dataItems[pointeeKey]
+	if !dataItemExists {
+		return writeTokens(enc,
+			notCapturedReason,
+			notCapturedReasonDepth,
+		)
+	}
+	address = pointedValue.Header().Address
+	goKind, ok := p.Pointee.GetGoKind()
+	if !ok {
+		return fmt.Errorf("no go kind for type %s (ID: %d)", p.Pointee.GetName(), p.Pointee.GetID())
+	}
+	if goKind != reflect.Pointer {
```

Comment:
If you're taking another pass, maybe worth adding commentary about why this is here